File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0: return 0
        if len(s) == 1: return 1

        maxLength, currLength = 1, 1
        l, r = 0, 1
        charSet = set(s[l])
        
        while r < len(s):
            if s[r] not in charSet:
                charSet.add(s[r])
                currLength += 1
                r += 1

            elif r - l > 1:
                charSet.remove(s[l])
                currLength -= 1
                l += 1

            else:
                l = r
                r += 1
            
            maxLength = max(currLength, maxLength)
        
        return maxLength

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
static const bool Booster = [](){
    std::ios_base::sync_with_stdio(false), std::cout.tie(nullptr), std::cin.tie(nullptr);
    return true;
}();
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> charSet;
        int l = 0, r = 0;
        int result = 0;
        while (r < s.size()) {
            while (charSet.contains(s[r])) {
                charSet.erase(s[l]);
                ++l;
            }
            charSet.insert(s[r]);
            ++r;
            result = max(result, r - l);
        }
        return result;
    }
};